microscopium/preprocess.py

```diff
@@ -704,3 +712,140 @@ def correct_image_illumination(im, illum, stretch_quantile=0, mask=None):
+    order = np.array(order)
+
+    # in case stream is passed, take one sip at a time ;)
+    ims = list(ims)
+    rows, cols = ims[0].shape[:2]
+    mrows, mcols = order.shape
+
+    montaged = np.zeros((rows * mrows, cols * mcols) + ims[0].shape[2:],
+                        dtype=ims[0].dtype)
+    for i in range(mrows):
+        for j in range(mcols):
+            montaged[rows*i:rows*(i+1), cols*j:cols*(j+1)] = ims[order[i, j]]
+    return montaged
+
+
+@tlz.curry
+def reorder(index_list, list_to_reorder):
+    """Curried function to reorder a list according to input indices.
+
+    Parameters
+    ----------
+    index_list : list of int
+        The list of indices indicating where to put each element in the
+        input list.
+    list_to_reorder : list
+        The list being reordered.
+
+    Returns
+    -------
+    reordered_list : list
+        The reordered list.
+
+    Examples
+    --------
+    >>> list1 = ['foo', 'bar', 'baz']
+    >>> reorder([2, 0, 1], list1)
+    ['baz', 'foo', 'bar']
+    """
+    return [list_to_reorder[j] for j in index_list]
+
+
+def montage_stream(ims, montage_order=None, channel_order=[0, 1, 2]):
+    """From a sequence of single-channel field images, montage multichannels.
+
+    Suppose the input is a list:
+
+    ```
+    ims = [green1, blue1, red1, green2, blue2, red2]
+    ```
+
+    with channel order:
+
+    ```
+    channel_order = [2, 0, 1]
+    ```
+
+    Then the output will be:
+
+    ```
+    [rgb1, rgb2]
+    ```
+
+    (The order of channels in the list is arbitrary; the default is
+    based on the data for which this software was created.)
+
+    Parameters
+    ----------
+    ims : iterator of array, shape (M, N)
+        A list of images in which consecutive images represent single
+        channels of the same image. (See example.)
+    order : list of int, optional
+        The order in which the channels appear.
+
+    Returns
+    -------
+    montaged_stream : iterator of arrays
+        An iterator of the images composed into multi-channel montages.
+
+    Examples
+    --------
+    >>> images = (i * np.ones((4, 5), dtype=np.uint8) for i in range(24))
+    >>> montaged = list(montage_stream(images, [[0, 1], [2, 3]], [2, 0, 1]))
+    >>> len(montaged)
+    2
+    >>> montaged[0].shape
+    (8, 10, 3)
+    >>> montaged[0][0, 0, :]
+    array([2, 0, 1], dtype=uint8)
+    >>> montaged[0][4, 5, :]
+    array([11,  9, 10], dtype=uint8)
+    >>> montaged[1][4, 5, :]
+    array([23, 21, 22], dtype=uint8)
+    """
+    if montage_order is None:
+        from .screens import cellomics
+        montage_order = cellomics.SPIRAL_CLOCKWISE_RIGHT_25
+    montage_order = np.array(montage_order)
+    ntiles = montage_order.size
+    nchannels = len(channel_order)
+    montage_ = fun.partial(montage, order=montage_order)
+    return tlz.pipe(ims, curried.partition(nchannels),
+                         curried.map(reorder(channel_order)),
+                         curried.map(np.dstack),
```

Comment:
Will this be able to handle image datasets where there's only 2 channels (eg the Cellomics datasets)? I had to write the ``stack_channels`` function instead of just using ``np.dstack`` so the missing channel would be replaced with all-zero images.